recipes/dns_config.rb

```diff
@@ -0,0 +1,148 @@
+    # Configure resolved to automatically append Route53 search domain in resolv.conf.
+    # On Ubuntu18 resolv.conf is managed by systemd-resolved.
+    # TODO verify if Ubuntu Desktop uses NetworkManager and we have to do a custom configuration for custom amis using this OS
+    replace_or_add "Append Route53 search domain in /etc/systemd/resolved.conf" do
+      path "/etc/systemd/resolved.conf"
+      pattern "Domains=*"
+      line "Domains=#{node['cfncluster']['cfn_dns_domain']}"
+    end
+  else
+    Chef::Log.info("Appending search domain '#{node['cfncluster']['cfn_dns_domain']}' to /etc/dhcp/dhclient.conf")
+    # Configure dhclient to automatically append Route53 search domain in resolv.conf
+    # - on CentOS6 resolv.conf is managed by network + dhclient,
+    # - on CentOS7, Alinux and Alinux2 by NetworkManager + dhclient,
+    # - on Ubuntu16 by networking + dhclient
+    replace_or_add "Append Route53 search domain in /etc/dhcp/dhclient.conf" do
+      path "/etc/dhcp/dhclient.conf"
+      pattern "append domain-name*"
+      line "append domain-name \" #{node['cfncluster']['cfn_dns_domain']}\";"
+    end
+
+    if platform?('centos') && node['platform_version'].to_i < 7
+      # On CentOS6 there is a dhclient configuration file for eth0
+      replace_or_add "Append Route53 search domain in /etc/dhcp/dhclient-eth0.conf" do
+        path "/etc/dhcp/dhclient-eth0.conf"
+        pattern "append domain-name*"
+        line "append domain-name \" #{node['cfncluster']['cfn_dns_domain']}\";"
+      end
+    end
+  end
+end
+
+# Utility method to restart network service according to the OS.
+def restart_network_service
+  network_service_name = value_for_platform(
+    ['centos'] => {
+      '<7.0' => 'network'
+    },
+    ['amazon'] => {
+      '>=2013' => 'network' # Alinux1
+    },
+    %w[ubuntu debian] => {
+      '16.04' => 'networking',
+      '>=18.04' => 'systemd-resolved'
+    },
+    'default' => 'NetworkManager'
+  )
+  Chef::Log.info("Restarting '#{network_service_name}' service, for platform #{node['platform']} '#{node['platform_version']}'")
+  service network_service_name.to_s do
+    action %i[restart]
+  end
+end
+
+if node['cfncluster']['cfn_scheduler'] == 'slurm'
+  # Heterogeneous Instance Type
+  if node['cfncluster']['cfn_dns_domain'].nil? ||
+     node['cfncluster']['cfn_dns_domain'].empty? ||
+     node['cfncluster']['cfn_dns_domain'] == "NONE"
+    Chef::Log.info("DNS domain not configured")
+    return
+  end
+
+  append_cluster_dns_domain
+
+  restart_network_service
+
+  if node['cfncluster']['cfn_node_type'] == "ComputeFleet"
+    # For compute node retrieve assigned hostname from DynamoDB and configure it
+    # - hostname: $QUEUE-static-$INSTANCE_TYPE_1-[1-$MIN1]
+    # - fqdn: $QUEUE-static-$INSTANCE_TYPE_1-[1-$MIN1].$CLUSTER_NAME.pcluster
+    Chef::Log.info("Retrieving assigned hostname")
+    begin
+      retries ||= 0
```

Comment:
to simplify retries you can use a ruby_block

Reply:
I started using the `ruby_block` but there are issues passing the retrieved value.

I might use `run_state` but it can only be used between chef resources and the `configure_hostname` function cannot be used in a chef resource.

Reply:
Fixe by using ruby_blocks, helpers and passing `compile_time = false` to the `hostname` resource that by default it is executed a t compile time.